lib/jekyll/regenerator.rb

```diff
@@ -20,18 +20,14 @@ def initialize(site)
     def regenerate?(document)
       case document
       when Page
-        document.asset_file? || document.data['regenerate'] ||
-          source_modified_or_dest_missing?(
-            site.in_source_dir(document.relative_path), document.destination(@site.dest)
-          )
+        regenerate_page?(document)
       when Document
-        !document.write? || document.data['regenerate'] ||
-          source_modified_or_dest_missing?(
-            document.path, document.destination(@site.dest)
-          )
+        regenerate_document?(document)
       else
-        source_path = document.respond_to?(:path)        ? document.path                    : nil
-        dest_path   = document.respond_to?(:destination) ? document.destination(@site.dest) : nil
+        source_path = document.respond_to?(:path) ? document.path : nil
+        dest_path = if document.respond_to?(:destination)
+                      document.destination(@site.dest)
+                    end
```

Comment:
What was wrong with the previous two lines? Just too long?

It seems like your solution for `source_path` and for `dest_path` are two ways of doing the same thing.